src/algebra/direct_sum_group.lean

```diff
@@ -0,0 +1,119 @@
+/-
+Copyright (c) 2019 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+
+Direct sum of abelian groups, indexed by a discrete type.
+-/
+import algebra.direct_sum
+
+open_locale big_operators
+
+universes u v w u₁
+
+namespace direct_sum
+
+variables {ι : Type v} [decidable_eq ι]
+variables {β : ι → Type w} [Π i, add_comm_group (β i)]
+
+instance : add_comm_group (direct_sum ι β) :=
+dfinsupp.add_comm_group
+
+instance mk.is_add_group_hom (s : finset ι) : is_add_group_hom (mk β s) :=
+{ map_add := λ _ _, dfinsupp.mk_add }
+
+@[simp] lemma mk_neg (s : finset ι) (x) : mk β s (-x) = -mk β s x :=
+is_add_group_hom.map_neg _ x
+
+@[simp] lemma mk_sub (s : finset ι) (x y) : mk β s (x - y) = mk β s x - mk β s y :=
+is_add_group_hom.map_sub _ x y
+
+instance of.is_add_group_hom (i : ι) : is_add_group_hom (of β i) :=
+{ map_add := λ _ _, dfinsupp.single_add }
+
+@[simp] lemma of_neg (i : ι) (x) : of β i (-x) = -of β i x :=
+is_add_group_hom.map_neg _ x
+
+@[simp] lemma of_sub (i : ι) (x y) : of β i (x - y) = of β i x - of β i y :=
+is_add_group_hom.map_sub _ x y
+
+variables {γ : Type u₁} [add_comm_group γ]
+variables (φ : Π i, β i → γ) [Π i, is_add_group_hom (φ i)]
+
+variables (φ)
+def to_group (f : direct_sum ι β) : γ := @to_add_monoid _ _ _ _ _ _ φ (λ i, by apply_instance) f
+variables {φ}
+
+-- TODO: reuse to_add_monoid.is_add_monoid_hom
+instance to_group.is_add_group_hom : is_add_group_hom (to_group φ) :=
+{ map_add := assume f g,
+begin
+  refine quotient.induction_on f (λ x, _),
+  refine quotient.induction_on g (λ y, _),
+  change ∑ i in _, _ = (∑ i in _, _) + (∑ i in _, _),
+  simp only, conv { to_lhs, congr, skip, funext, rw is_add_hom.map_add (φ i) },
+  simp only [finset.sum_add_distrib],
+  congr' 1,
+  { refine (finset.sum_subset _ _).symm,
+    { intro i, simp only [multiset.mem_to_finset, multiset.mem_add], exact or.inl },
+    { intros i H1 H2, simp only [multiset.mem_to_finset, multiset.mem_add] at H2,
+      rw [(x.3 i).resolve_left H2, is_add_monoid_hom.map_zero (φ i)] } },
+  { refine (finset.sum_subset _ _).symm,
+    { intro i, simp only [multiset.mem_to_finset, multiset.mem_add], exact or.inr },
+    { intros i H1 H2, simp only [multiset.mem_to_finset, multiset.mem_add] at H2,
+      rw [(y.3 i).resolve_left H2, is_add_monoid_hom.map_zero (φ i)] } }
+end }
```

Comment:
The body of this tactic is identical to `to_add_monoid.is_add_monoid_hom`, but I can't work out how to reference it.

Reply:
I think you should be able to write `.. to_add_monoid.is_add_monoid_hom` at the end of your structure.

On the other hand, we might want to "bundle" the morphisms right from the start.